src/main/scala/cromwell/engine/workflow/WorkflowActor.scala

```diff
@@ -196,129 +239,156 @@ case class WorkflowActor(workflow: WorkflowDescriptor, backend: Backend)
+      // Collector keys don't have to be in Running state.
+      val updatedData = data.addPersisting(collectorKey)
+      callCompletedWhileAborting(collectorKey, outputs, returnCode)
+    case Event(CallAborted(callKey), data) if callKey.isPersistedRunning(data) =>
+      persistStatus(callKey, ExecutionStatus.Aborted, None)
+      val updatedData = data.addPersisting(callKey)
+      (if (isWorkflowAborted) transition(WorkflowAborted) else stay()) using updatedData
+    case Event(CallFailed(callKey, returnCode, failure), data) if callKey.isPersistedRunning(data) =>
+      persistStatus(callKey, ExecutionStatus.Failed, returnCode)
+      val updatedData = data.addPersisting(callKey)
+      (if (isWorkflowAborted) transition(WorkflowAborted) else stay()) using updatedData
+    case Event(m, _) =>
       logger.error("Unexpected message in Aborting state: " + m.getClass.getSimpleName)
-      if (isWorkflowAborted) goto(WorkflowAborted) using NoFailureMessage else stay()
+      if (isWorkflowAborted) transition(WorkflowAborted) else stay()
   }
 
-  when(WorkflowAborted)(FSM.NullFunction)
-
   whenUnhandled {
     case Event(AbortWorkflow, _) =>
       context.children foreach { _ ! CallActor.AbortCall }
-      goto(WorkflowAborting) using NoFailureMessage
+      transition(WorkflowAborting)
+    case Event(callMessage: CallMessage, _) =>
+      logger.debug(s"Dropping message for ineligible key: ${callMessage.callKey.tag}")
+      // Running and Aborting have explicit handlers for Running and no pending writes, so either we are not in those
+      // states or the specified key is no longer Running and/or has pending writes.
+      stay()
+    case Event(PersistenceCompleted(callKey), data) =>
+      logger.debug(s"Got PersistenceCompleted(${callKey.tag})")
+      val updatedData = data.removePersisting(callKey)
+      stay using updatedData
+    case Event(PerformTransition(toState), data) if data.pendingExecutions.isEmpty && !stateName.isTerminal =>
+      // stateName is the *current* state, from which there should not be transitions.
+      logger.info(s"transitioning from $stateName to $toState.")
+      goto(toState)
+    case Event(PerformTransition(toState), _) if stateName.isTerminal =>
+      // Drop the message.  This is a race condition with a retried PerformTransition message when
+      // the FSM has transitioned to a terminal state between the original and retried messages.
+      stay()
+    case Event(message @ PerformTransition(toState), _) =>
+      logger.debug(s"Rescheduling transition request to $toState due to pending persists")
+      context.system.scheduler.scheduleOnce(100 milliseconds, self, message)
+      stay()
+    case Event(AsyncFailure(t), data) if data.pendingExecutions.isEmpty =>
+      logger.error(t.getMessage, t)
+      transition(WorkflowFailed)
+    case Event(message @ AsyncFailure(t), _) =>
+      logger.warn("Rescheduling AsyncFailure due to pending execution writes")
```

Comment:
If the message is eventually logged, perhaps this should be `logger.debug`? Otherwise I'd expect this block to output multiple warning messages every 100ms with the equivalent of "Wait for it... you're about to see what the actual failure is!"